x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/grouping/Histogram.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.sql.expression.function.grouping;
+
+import org.elasticsearch.xpack.sql.expression.Expression;
+import org.elasticsearch.xpack.sql.expression.Expressions;
+import org.elasticsearch.xpack.sql.expression.Expressions.ParamOrdinal;
+import org.elasticsearch.xpack.sql.expression.Literal;
+import org.elasticsearch.xpack.sql.tree.Location;
+import org.elasticsearch.xpack.sql.tree.NodeInfo;
+import org.elasticsearch.xpack.sql.type.DataType;
+import org.elasticsearch.xpack.sql.type.DataTypes;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.TimeZone;
+
+public class Histogram extends GroupingFunction {
+
+    private final Literal interval;
+    private final TimeZone timeZone;
+
+    public Histogram(Location location, Expression field, Expression interval, TimeZone timeZone) {
+        super(location, field);
+        this.interval = (Literal) interval;
+        this.timeZone = timeZone;
+    }
+
+    public Literal interval() {
+        return interval;
+    }
+
+    public TimeZone timeZone() {
+        return timeZone;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        TypeResolution resolution = Expressions.typeMustBeNumericOrDate(field(), "HISTOGRAM", ParamOrdinal.FIRST);
+        if (resolution == TypeResolution.TYPE_RESOLVED) {
+            // interval must be Literal interval
+            if (field().dataType() == DataType.DATE) {
+                resolution = Expressions.typeMustBe(interval, DataTypes::isInterval, "(Date) HISTOGRAM", ParamOrdinal.SECOND, "interval");
+            } else {
+                resolution = Expressions.typeMustBeNumeric(interval, "(Numeric) HISTOGRAM", ParamOrdinal.SECOND);
+            }
+        }
+
+        return resolution;
+    }
+
+
+    @Override
+    public DataType dataType() {
+        return field().dataType();
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
```

Comment:
Why the `replaceChildren(List<Expression>)` here and not defined as `final` in `GroupingFunction` enforcing the replacement of a single children? `GroupingFunction` would then define a `protected abstract GroupingFunction replaceChildren(Expression newExpression);`. Are we expecting more grouping functions with variable list of children in the future?

Reply:
We don't know what the future will bring - the pattern above is used for `Unary[..]Function`s which a grouping function might be, but we don't know yet.
Adding it now results in 2 methods without any reuse so it's arguable whether it makes sense to do it or not. In general, I prefer to add a pattern when it applies to at least 2 instances, otherwise it's more of a premature recognition.
I've added it nevertheless.